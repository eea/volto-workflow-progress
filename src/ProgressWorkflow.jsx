import React, { useEffect, useState } from 'react';
import { Portal } from 'react-portal';
import { useDispatch, useSelector } from 'react-redux';
import { getWorkflowProgress } from './actions';
import './less/editor.less';

/**
 * @summary The React component that shows progress tracking of selected content.
 */
const ProgressWorkflow = (props) => {
  const { content, pathname } = props;
  const currentStateKey = content?.review_state;
  const dispatch = useDispatch();

  const [visible, setVisible] = useState(false);
  const [isToolbarOpen, setIsToolbarOpen] = useState(false);
  const [workflowProgressSteps, setWorkflowProgressSteps] = useState([]);
  const [currentState, setCurrentState] = useState(null);
  const workflowProgress = useSelector((state) => state?.workflowProgress);
  const thestate = useSelector((state) => state);

  // toggle progress component visible by clicking on the button
  const setVisibleSide = () => {
    setVisible(!visible);
  };

  useEffect(() => {
    const findCurrentState = (steps, done) => {
      const arrayContainingCurrentState = steps.find(
        (itemElements) => itemElements[1] === done,
      );
      const indexOfCurrentSateKey = arrayContainingCurrentState[0].indexOf(
        currentStateKey,
      );
      const title = arrayContainingCurrentState[2][indexOfCurrentSateKey];
      const description = arrayContainingCurrentState[3][indexOfCurrentSateKey];

      setCurrentState({
        done,
        title,
        description,
      });
    };
    const hasToolbar = document.querySelector('#toolbar .toolbar-actions');

    setIsToolbarOpen(!!hasToolbar);
    // filter out paths that don't have workflow (home, login etc)
    if (
      workflowProgress.result &&
      !workflowProgress.workflow?.error &&
      Array.isArray(workflowProgress.result.steps)
    ) {
      findCurrentState(
        workflowProgress.result.steps,
        workflowProgress.result.done,
      );
      setWorkflowProgressSteps(workflowProgress.result.steps);
    } else {
      setCurrentState(null); // reset current state only if a path without workflow is
      // chosen to avoid flicker for those that have workflow
    }
  }, [currentStateKey, workflowProgress]);

  useEffect(() => {
    dispatch(getWorkflowProgress(pathname)); // the are paths that don't have workflow (home, login etc)
  }, [dispatch, pathname]);

  const itemTracker = (tracker) => (
    <li
      key={`progress__item ${tracker[0]}`}
      className={`progress__item ${
        tracker[0].indexOf(currentStateKey) > -1
          ? 'progress__item--active'
          : tracker[1] < currentState.done
          ? 'progress__item--completed'
          : 'progress__item--next'
      }`}
    >
      {tracker[2].map((title, index) => (
        <p
          key={`progress__title ${tracker[0]}${index}`}
          className={`progress__title ${
            currentState.title !== title ? 'title-incomplete' : ''
          }`}
        >
          {title}
        </p>
      ))}
    </li>
  );

  const currentStateClass = {
    published: 'published',
    private: 'private',
  };

  return (
    isToolbarOpen && (
      <Portal
        node={__CLIENT__ && document.querySelector('#toolbar .toolbar-actions')}
      >
        {currentState ? (
          <>
            <div>
              <button
                className={`circle-right-btn ${
                  currentStateClass[currentStateKey]
                    ? `review-state-${currentStateKey}`
                    : ''
                }`}
                id="toolbar-cut-blocks"
                onClick={setVisibleSide}
              >
                {`${currentState.done}%`}
              </button>
              {visible ? (
                <div className="sidenav-ol">
                  <ul className="progress">
                    {workflowProgressSteps.map((progressItem) =>
                      itemTracker(progressItem),
                    )}
                  </ul>
                </div>
              ) : null}
            </div>
            <div
              className={`review-state-text ${
                currentStateClass[currentStateKey]
                  ? `review-state-${currentStateKey}`
                  : ''
              }`}
            >
              {currentState.title}
            </div>
          </>
        ) : (
          <></>
        )}
      </Portal>
    )
  );
};
export default ProgressWorkflow;
