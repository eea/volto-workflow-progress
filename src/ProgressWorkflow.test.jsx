// âœ“ ca nu crapa la obiect gol
// daca vine cum trebuie vedem procent, titlu, sidebar progress tracker
// ca nu face nimic daca nu e pe ce ruta trebuie
// ca nu face nimic daca nu e logat userul
// ca nu face nimic daca nu exista toolbar

import React from 'react';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import ProgressWorkflow from './ProgressWorkflow';

const mockStore = configureStore();
const props = {
  content: {
    '@id': 'http://localhost:3000/api/',
    review_state: 'pending',
  },
  pathname: '/',
};
const propsEmpty = {};

describe('ProgressWorkflow', () => {
  it('renders the ProgressWorkflow component without breaking if props and workflow are empty', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      workflowProgress: {},
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <ProgressWorkflow
            pathname="/test"
            {...propsEmpty}
            hasToolbar={true}
          />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders the ProgressWorkflow component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      workflowProgress: {
        '@id': 'http://localhost:3000/api/my-page/@workflow.progress',
        done: 50,
        steps: [
          [
            ['private'],
            25,
            ['Private'],
            ['Can only be seen and edited by the owner.'],
          ],
          [
            ['pending'],
            50,
            ['Pending review'],
            ['Waiting to be reviewed, not editable by the owner.'],
          ],
          [
            ['review_one', 'review_two', 'review_three', 'review_four'],
            75,
            [
              'Review One: Technical',
              'Review two: Head of Technical',
              'Review three: Head of Department',
              'Review Four: CTO',
            ],
            [
              'Review One: Technical',
              'Review by Head of Tech Depart',
              'Review by Head of Department',
              'Review by CTO',
            ],
          ],
          [
            ['published', 'visible'],
            100,
            ['Published', 'Public draft'],
            [
              'Visible to everyone, not editable by the owner.',
              'Visible to everyone, but not approved by the reviewers.',
            ],
          ],
        ],
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <ProgressWorkflow pathname="/test" {...props} hasToolbar={true} />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders the ProgressWorkflow component with Percent showing correct value', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      workflowProgress: {
        '@id': 'http://localhost:3000/api/my-page/@workflow.progress',
        done: 50,
        steps: [
          [
            ['private'],
            25,
            ['Private'],
            ['Can only be seen and edited by the owner.'],
          ],
          [
            ['pending'],
            50,
            ['Pending review'],
            ['Waiting to be reviewed, not editable by the owner.'],
          ],
          [
            ['review_one', 'review_two', 'review_three', 'review_four'],
            75,
            [
              'Review One: Technical',
              'Review two: Head of Technical',
              'Review three: Head of Department',
              'Review Four: CTO',
            ],
            [
              'Review One: Technical',
              'Review by Head of Tech Depart',
              'Review by Head of Department',
              'Review by CTO',
            ],
          ],
          [
            ['published', 'visible'],
            100,
            ['Published', 'Public draft'],
            [
              'Visible to everyone, not editable by the owner.',
              'Visible to everyone, but not approved by the reviewers.',
            ],
          ],
        ],
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <ProgressWorkflow pathname="/test" {...props} hasToolbar={true} />
        </MemoryRouter>
      </Provider>,
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
