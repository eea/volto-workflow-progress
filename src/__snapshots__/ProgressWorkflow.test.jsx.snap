// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressWorkflow renders the ProgressWorkflow component 1`] = `null`;

exports[`ProgressWorkflow renders the ProgressWorkflow component with Percent showing correct value 1`] = `null`;

exports[`ProgressWorkflow renders the ProgressWorkflow component without breaking if props and workflow are empty 1`] = `null`;
